### PR TITLE
User cancels submission before API accepts it

### PR DESCRIPTION
**Pull request by kntsoriano**
https://github.com/exercism/v3-project-management/issues/53

## Description
This PR adds a "Cancel" button to allow the user to cancel their submission while a response from the submissions API isn't received.

## Questions
1. @iHiD How does this look like from the backend? Is it safe to assume that we can just cancel the request if the API hasn't responded in time? I'm probably thinking that it isn't, so we might need to change the API call to not take too long, and just acknowledge that it has received the request and respond right away? The current API is already queueing jobs and other things before it responds. I decided to push through with this, though, as this solves the instance where the user might have internet problems.

### CODE REVIEW

**Comment by iHiD** [comment]
Could we have it so that once the submission comes back, it immediately gets cancelled. (We could keep this as an async function while continuing the "normal" codepath)?

**Comment by kntsoriano** [comment]
Could you explain further?

**Comment by iHiD** [comment]
So:
1. We create a request
2. The user wants to cancel the request.

We now want to tell the server the submission is cancelled. However, we don't have a submissionId so we can't cancel it (unless we add an extra corrolation token, which I'm a little meh about). So the easiest option would be to wait for the request to return, then when we get the response with the token, immediately send another request to cancel it.

So rather than just throwing away the request and imagining it never happened, we wait for it to come back and send another request to cancel it. This could just be a background thread - the UI doesn't have to wait for this to happen.

There's still a small chance that some requests get through but we drop the response from the server, but I'm ok with that tradeoff.

**Comment by iHiD** [comment]
It's also worth noting that the actual API request is really fast. The only reason it's slow atm is that it does a Git Fetch from GitHub in development mode, which is very slow.

**Comment by iHiD** [comment]
You know what, ignore me. I'm ok with this. The chance of someone cancelling within the 1s of starting is small, so this is fine.

**Comment by iHiD** [comment]
@SleeplessByte @ErikSchierboom Could you both take a look at this one pls.

**Review by SleeplessByte — COMMENTED** [review]
I need to respond to this. Please ping me if I haven't when y'all have discussed it and want to merge it in.

Few consistency (react related) issues I see -- I'll point them out then.

File: app/javascript/components/student/Editor.tsx [hunk 1]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
+  const [{ submission, isSubmitting, code }, dispatch] = useReducer(reducer, {
+    submission: undefined,
+    isSubmitting: false,
+    code: '',
+  })
   const controllerRef = useRef(new AbortController())
+  const abortFetch = useCallback(() => {
+    controllerRef.current.abort()
+    controllerRef.current = new AbortController()
+  }, [controllerRef.current])
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:72]:
```suggestion
  }, [controllerRef])
```

This callback reaches _into_ the `controllerRef`. Its dependency should be the ref, not the current value. _Otherwise_ you will get aborts when for ANY other reason this component re-renders.

File: app/javascript/components/student/Editor.tsx [hunk 2]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
 
   useEffect(() => {
     return () => {
-      controllerRef.current.abort()
+      abortFetch()
     }
   }, [])
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:78]:
```suggestion
  }, [abortFetch])
```

File: app/javascript/components/student/Editor.tsx [hunk 3]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
 
-  function submit(code: string) {
-    controllerRef.current.abort()
-    controllerRef.current = new AbortController()
-    setSubmission(undefined)
+  useEffect(() => {
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:80]: I find it hard to follow why we need all the different effects. So apparently `isSubmitting` can change on its own. And then you want to abort? Why? It feels like you only want that because you want a `new AbortController`.

File: app/javascript/components/student/Editor.tsx [hunk 4]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
     fetchJSON(endpoint, {
       method: 'POST',
       signal: controllerRef.current.signal,
       body: JSON.stringify({ files: { file: code } }),
-    }).then((json: any) => {
-      setSubmission(typecheck<Submission>(json, 'submission'))
     })
-  }
+      .then((json: any) => {
+        dispatch({
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:103]: Should check if still mounted. The abort controller takes care of this, but we'd really want to guard for it if the abortcontroller fails (because a browser is stupid). It's also good practice to think about it)

File: app/javascript/components/student/Editor.tsx [hunk 5]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
   endpoint: string
   timeout?: number
 }) {
-  const [submission, setSubmission] = useState<Submission>()
+  const [{ submission, isSubmitting, code }, dispatch] = useReducer(reducer, {
+    submission: undefined,
+    isSubmitting: false,
+    code: '',
+  })
   const controllerRef = useRef(new AbortController())
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:68]:
```suggestion
  const controllerRef = useRef<undefined | AbortController>()
```

File: app/javascript/components/student/Editor.tsx [hunk 6]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
+  const abortFetch = useCallback(() => {
+    controllerRef.current.abort()
+    controllerRef.current = new AbortController()
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:71]:
```suggestion
```

File: app/javascript/components/student/Editor.tsx [hunk 7]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
+  }, [isSubmitting])
+
+  useEffect(() => {
+    if (code === '') {
+      return
+    }
+
+    if (isSubmitting) {
+      abortFetch()
+    }
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:95]:
```suggestion
    if (abortController.current) {
      abortController.current.abort()
      abortController.current = new AbortController()
    }
```

File: app/javascript/components/student/Editor.tsx [hunk 8]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
+        dispatch({
+          type: 'submitted',
+          payload: { submission: typecheck<Submission>(json, 'submission') },
+        })
+      })
+      .catch((responseOrError) => {
+        if (responseOrError.name === 'AbortError') {
+          console.log('Fetch cancelled')
+        }
+      })
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:112]:
```suggestion
      }).finally(() => { abortController.current = undefined }))
```
You probably need to polyfill this -- webpack should take care of this.

File: app/javascript/components/student/Editor.tsx [hunk 9]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
+      .then((json: any) => {
+        dispatch({
+          type: 'submitted',
+          payload: { submission: typecheck<Submission>(json, 'submission') },
+        })
+      })
+      .catch((responseOrError) => {
+        if (responseOrError.name === 'AbortError') {
+          console.log('Fetch cancelled')
+        }
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:111]:
```suggestion
        }
        
        throw responseOrError
```

Rethrow, or this error is lost forever.

File: app/javascript/components/student/Editor.tsx [hunk 10]
```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
...
-      setSubmission(typecheck<Submission>(json, 'submission'))
     })
-  }
+      .then((json: any) => {
+        dispatch({
+          type: 'submitted',
+          payload: { submission: typecheck<Submission>(json, 'submission') },
+        })
+      })
+      .catch((responseOrError) => {
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:108]: Can this really be a response or error?

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 11]
```diff
@@ -70,6 +70,7 @@ export function TestRunSummary({
     tests: [],
   })
   const haveTestsResolved = RESOLVED_TEST_STATUSES.includes(testRun.status)
+  const timer = useRef<number>()
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/TestRunSummary.tsx:73]:
```suggestion
  const timer = useRef<number | undefined>()
```

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 12]
```diff
@@ -86,9 +87,18 @@ export function TestRunSummary({
...
+      () => dispatch({ type: 'testRun.timeout' }),
+      timeout
+    )
   }, [testRun.status])
 
+  useEffect(() => {
+    return () => {
+      clearTimeout(timer.current)
+    }
+  }, [])
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/TestRunSummary.tsx:100]:
```suggestion
  }, [timer])
```

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 13]
```diff
@@ -86,9 +87,18 @@ export function TestRunSummary({
       return
     }
 
-    setTimeout(() => dispatch({ type: 'testRun.timeout' }), timeout)
+    timer.current = window.setTimeout(
+      () => dispatch({ type: 'testRun.timeout' }),
+      timeout
+    )
   }, [testRun.status])
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/TestRunSummary.tsx:94]:
```suggestion
  }, [timer, testRun.status])
```

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 14]
```diff
@@ -86,9 +87,18 @@ export function TestRunSummary({
       return
     }
 
-    setTimeout(() => dispatch({ type: 'testRun.timeout' }), timeout)
+    timer.current = window.setTimeout(
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/TestRunSummary.tsx:90]:
```suggestion
    timer.current = setTimeout(
```

File: app/javascript/utils/fetch-json.ts [hunk 15]
```diff
@@ -5,23 +5,28 @@ export async function fetchJSON(endpoint: string, options: any) {
...
+      !contentType ||
+      (!contentType.includes('+json') &&
+        !contentType.includes('application/json'))
+    ) {
+      throw response
+    }
+
+    return camelizeKeys(await response.json())
+  } catch (responseOrError) {
+    return Promise.reject(responseOrError)
```
Comment (SleeplessByte) [inline app/javascript/utils/fetch-json.ts:30]: Errors thrown in `async` functions are automatically turned into promise. Remove the entire `try` `catch`:

```typescript
async function hello(boom = false) {
  if (boom) {
    throw new Error('boom')
  }

  return 42
}

hello().then(console.log)
// => 42

hello().catch(console.error)
// => Error('boom')
```

**Comment by SleeplessByte** [comment]
So. @kntsoriano , see the review above, but I actually have a different way to solve this.

```tsx
export function Editor({
  endpoint,
  timeout = 60000,
}: {
  endpoint: string;
  timeout?: number;
}) {
  const [submission, setSubmission] = useState<Submission>();
  const [{ submission, isSubmitting, code }, dispatch] = useReducer(reducer, {
    submission: undefined,
    isSubmitting: false,
    code: '',
  });

  const abortControllerRef = useRef<AbortController | undefined>();
  const abort = useCallback(() => {
    abortControllerRef.current?.abort();
    abortControllerRef.current = undefined;
  }, [abortControllerRef]);

  const submit = useCallback(
    (nextCode: string) => {
      if (nextCode.trim().length === 0) {
        return;
      }

      abort();
      abortControllerRef.current = new AbortController();

      dispatch({ type: 'submitting' });

      fetch(/* ... */)
        .then(() => {
          /* ... */
          dispatch({ type: 'submitted' });
        })
        .catch(() => {
          /* ... */
          dispatch({ type: 'cancelled' }); // it actually is! you're missing this
        })
        .finally(() => {
          abortControllerRef.current = undefined;
        });
    },
    [abort, abortControllerRef]
  );

  const cancel = useCallback(() => {
    abort();
    dispatch({ type: 'cancelled' });
  }, [dispatch, abort]);

  // Abort ongoing requests when this component unmounts
  useEffect(() => {
    return abort;
  }, [abort])
  
  // pass down cancel and submit
  return (
    <div>
      <CodeEditor onSubmit={submit} onCancel={cancel} />
      {isSubmitting && <Submitting onCancel={cancel}/>}
    </div>
  )
}
```

What this does is create callbacks for the actual actions. It allows you to have side-effects _before_ you dispatch. Makes it much easier to follow the code and you don't need all the useEffects. 

[`.finally`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally) might not exist; you can add it via `tsconfig.json`, or add the line `abortControllerRef.current = undefined;` twice (once in the `then`, once in the `catch`)

**Comment by iHiD** [comment]
@kntsoriano I much prefer the callbacks syntax too. I've read the React docs and they recommend the dispatch thing (so your version is "correct"), but I think it's really hard to get your head around, compared to the callbacks. And it feels really leaky to be passing the dispatch around.

So I'd rather we use the callbacks approach if you're happy with that? (cc @ErikSchierboom for discussion too!)

---

(Thank you @SleeplessByte!)

**Comment by SleeplessByte** [comment]
For the timer, I would extract a custom hook:

```typescript
function useTimeout(timeout: number, stableCallback: () => void) {
  const timerRef = useRef<number | undefined>()
  const onTimeout = useCallback(() => {
    stableCallback()
    timerRef.current = undefined
  }, [stableCallback, timerRef])

  useEffect(() => {
    timerRef.current = setTimeout(stableCallback, onTimeout)
  }, [onTimeout, stableCallback])

  useEffect(() => {
    // Cancel when hook unmounts
    return () => {
      clearTimeout(timerRef.current)
    }
  }, [timerRef])
}
```

However, this hook starts "running" as soon as you mount it. You can add another variable to the hook to make it "conditional", you could add a `start` and `cancel` callback that is returned from the hook, so you can make it imperative (in your `useEffect`), but far far far easier, is to be declarative:

```tsx
function TimeoutTestRun({ timeout: number, onTimeout }: { timeout: number, onTimeout: () => void }) {
  onTimeout(timeout, onTimout)
  return null 
}

// ...
const cancel = useCallback(() => dispatch({ type: 'testRun.timeout' }), [dispatch])

return (
    <div>
      {hasTestBeenResolved ? null : <TimeoutTestRun timeout={timeout} onTimout={cancel} />}
      <p>Status: {testRun.status}</p>
      {/*...*/}
```

**Comment by kntsoriano** [comment]
@iHiD @SleeplessByte Cool. I will rewrite this to use callbacks.

File: app/javascript/components/student/editor/TestRunSummary.tsx (hunk 14 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/TestRunSummary.tsx:90]: I get this error: Type 'Timeout' is not assignable to type 'number'.

https://stackoverflow.com/questions/45802988/typescript-use-correct-version-of-settimeout-node-vs-window/50428377

What should I do?
